### PR TITLE
Initial branch

### PR DESCRIPTION
**Pull request by AikoChou**

### CODE REVIEW

File: config.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    profile = True,
+
+    stats_max_age_days = 90,
+
+    user_agent = 'citationdetective',
+
+)
+
+_CITATION_NEEDED_CONFIG = dict(
```
Comment (eggpi) [inline config.py]: The first three here only apply to English, right? So they could be part of the English-specific config, and `max_seq_length` could be global.

File: config.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    vocb_path = os.path.expanduser('~/citation-needed/embeddings/word_dict_en.pck'),
+
+    # Embeddings for the section titles
+    section_path = os.path.expanduser('~/citation-needed/embeddings/section_dict_en.pck'),
+
+    # Tensorflow models to detect Citation Need for English
+    model_path = os.path.expanduser('~/citation-needed/models/fa_en_model_rnn_attention_section.h5'),
+
+    # Maximum length of all sequences
+    max_seq_length = 187,
```
Comment (eggpi) [inline config.py]: What is a sequence? Maybe we can come up with a more obvious name?

File: config.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    archive_duration_days = 90,
+
+    # Where to put various logs
+    log_dir = os.path.join(os.path.expanduser('~'), 'logs'),
+
+    flagged_off = [],
+
+    profile = True,
+
+    stats_max_age_days = 90,
```
Comment (eggpi) [inline config.py]: I think none of the config options up to here actually apply to us now -- they're for controlling backups, profiling and stats (https://tools.wmflabs.org/citationhunt/stats.html).

File: config.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+# A base configuration that all languages "inherit" from.
+_BASE_LANG_CONFIG = dict(
+
+    statement_max_size = 200,
+
+    context_max_size = 800,
+    
+    hyperlink_regex = "http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+#]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+",
+
+    sentence_regex = '(?<!\w\.\w.)(?<![A-Z][a-z]\.)(?<=\.|\?)\s',
```
Comment (eggpi) [inline config.py]: How about we use https://www.nltk.org/api/nltk.tokenize.html#nltk.tokenize.sent_tokenize ?

File: parse.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            # store all (sub)heading names
+            headings = [h.title for h in section.filter_headings()]
+
+        # split section content into paragraphs
+        section_parag = re.split("\n+", section.strip_code())
+
+        for i, paragraph in enumerate(section_parag):
+            pid = mkid(section_name+str(i))
+            paragraphs[pid] = paragraph
+            # clean hyperlinks which strip_code() did not remove
```
Comment (eggpi) [inline parse.py]: Do you have an example where this happens? I think it should not happen in most cases according to the documentation for [strip_code](https://mwparserfromhell.readthedocs.io/en/latest/api/mwparserfromhell.html#mwparserfromhell.wikicode.Wikicode.strip_code), but we can also process links individually if we have to with [filter_external_links](https://mwparserfromhell.readthedocs.io/en/latest/api/mwparserfromhell.html#mwparserfromhell.wikicode.Wikicode.filter_external_links).

Also, we should check with @mirrys whether we really can't just keep the links. As I understand it, the model only has a fixed set of words that it uses for predictions, so anything not in that list (including links) will just be ignored later on anyway...

File: database.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    def connect_and_initialize():
+        db = _connect_to_wp_mysql(cfg)
+        with db.cursor() as cursor:
+            cursor.execute('USE ' + cfg.database)
+        return db
+    return _RetryingConnection(connect_and_initialize)
+
+# Methods for use in batch scripts, not the serving frontend. These set up the
+# databases, help populate the scratch database and swap it with the serving
+# database.
```
Comment (eggpi) [inline database.py]: Rephrase or remove this comment? We don't have a "serving frontend" so it's misleading.

File: database.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+def _make_tools_labs_dbname(cursor, database, lang_code):
+    cursor.execute("SELECT SUBSTRING_INDEX(USER(), '@', 1)")
+    user = cursor.fetchone()[0]
+    return '%s__%s_%s' % (user, database, lang_code)
+
+def _use(cursor, database, lang_code):
+    cursor.execute('USE %s' % _make_tools_labs_dbname(
+        cursor, database, lang_code))
+
+def init_scratch_db():
```
Comment (eggpi) [inline database.py]: You'll also want the function for installing the scratch database, right? https://github.com/eggpi/citationhunt/blob/master/chdb.py#L241

Basically in CH we create a "scratch" database, then atomically swap the real one with it, at which point it becomes the new real one that is used for serving.

File: print_pageids_from_wikipedia.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import config
+import database
+
+def print_pageids_from_wikipedia():
+    cfg = config.get_localized_config()
+    db = database.init_wp_replica_db(cfg.lang_code)
+    cursor = db.cursor()
+
+    cursor.execute(
+        'SELECT page_id FROM page where page_namespace=0 LIMIT %s' % int(cfg.num_pages * cfg.X * 0.01))
```
Comment (eggpi) [inline print_pageids_from_wikipedia.py]: Rather than hard-coding the number of pages, which will inevitable go out of date, how about something like:

```sql
SELECT page_id FROM page WHERE page_namespace = 0 AND RAND() < cfg.articles_sampling_fraction
```

http://bigdatums.net/2017/01/07/select-n-percent-records-mysql/

File: config.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            'Further reading',
+            'Notes',
+        ],
+        
+        # https://en.wikipedia.org/wiki/Special:Statistics
+        # As of 31 December 2019, there are 5,989,239 articles in the English Wikipedia.
+        num_pages = 6000000,
+
+        # X% of all articles
+        X = 1,
```
Comment (eggpi) [inline config.py]: How about calling this `articles_sampling_rate`, which is way more explicit?

File: utils.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    return s.encode('utf-8')
+
+def mkid(s):
+    return hashlib.sha1(e(s)).hexdigest()[:2*4]
+
+def running_in_tools_labs():
+    return (os.path.exists('/etc/wmflabs-project') or
+        os.path.exists('/etc/wmcs-project'))
+
+def text_to_word_list(text):
```
Comment (eggpi) [inline utils.py]: @mirrys Could this be replaced with https://www.nltk.org/api/nltk.tokenize.html#nltk.tokenize.punkt.PunktLanguageVars.word_tokenize instead?

File: parse.py (hunk 5 above)
Comment (AikoChou) [reply parse.py]: For example, the first paragraph in "[Business](https://en.wikipedia.org/wiki/Business)", strip_code doesn’t remove the content between \<ref\> and \</ref\> which possibly contains text and hyperlinks. But I just found I can use [filter_tags](https://mwparserfromhell.readthedocs.io/en/latest/api/mwparserfromhell.html#mwparserfromhell.wikicode.Wikicode.filter_tags) to catch them. 

The text returned by strip_code are the following which bold text should be removed: 

> Business is the activity of making one's living or making money by producing or buying and selling products (such as goods and services). **Compare: American Heritage Dictionary "business [:] 1. The activity of buying and selling  commodities, products, or services".** **Longman Business English Dictionary** **Longman Dictionary of Contemporary English "business [:] 1 [...] the activity of making money by producing or buying and selling goods, or providing services".** **Oxford Living Dictionaries "business [:] 2 The practice of making one's living by engaging in commerce."** Simply put, it is "any activity or enterprise entered into for profit. It does not mean it is a company, a corporation, partnership, or have any such formal organization, but it can range from a street peddler to General Motors."**Burton's Legal Thesaurus, 4E. S.v. "business." Retrieved April 1, 2018 from https://legal-dictionary.thefreedictionary.com/business** 

And I don't think we can keep these links and texts. Because for those words which are not in the list, the model fills `'UNK'` instead of ignoring them.

File: config.py (hunk 3 above)
Comment (AikoChou) [reply config.py]: Oh ok! I’ll remove them. I was not sure what are these for.

File: config.py (hunk 1 above)
Comment (AikoChou) [reply config.py]: Yes, the first three apply to English. I've checked `max_seq_length` also differs from language to language model, so I'll put them to English-specific config.

File: config.py (hunk 2 above)
Comment (AikoChou) [reply config.py]: `max_seq_length` is for padding sequences (or more precisely word vectors) to the same length so as to use as the input for the RNN model. (https://keras.io/preprocessing/sequence/#pad_sequences) 

Or we change the name to `word_vector_length`?

File: config.py (hunk 9 above)
Comment (AikoChou) [reply config.py]: Haha yeah that is better!

File: database.py (hunk 7 above)
Comment (AikoChou) [reply database.py]: Oh yes, I need the function as well.

File: database.py (hunk 6 above)
Comment (AikoChou) [reply database.py]: Ok!

File: config.py (hunk 2 above)
Comment (eggpi) [reply config.py]: I like `word_vector_length` :) Maybe also leave a quick explanation as a comment above? The current comment is not very helpful I think...

File: parse.py (hunk 5 above)
Comment (eggpi) [reply parse.py]: Ah yes, I guess it makes sense that `strip_code` doesn't remove tags.

But how does `'UNK'` affect the output of the model? Does their presence actually change the score for a sentence?

**Comment by eggpi** [comment]
Thanks! As you make changes in response to comments, please push them as other commits in this branch so I can see them in the same PR and continue the review!

File: parse.py (hunk 5 above)
Comment (AikoChou) [reply parse.py]: I think because the input for the RNN model is a fixed length of the word vectors, if we don't remove extra hyperlinks, they become many `UNK` filling the word vectors. When the length exceeds the limitation of maximum word vectors, they are truncated and some meaningful words are missing.

File: parse.py [hunk 11]
```diff
@@ -131,11 +144,12 @@ def parse(pageids):
             id = mkid(title+section+text)
             row = [id, text, paragraphs[pidx], section, revid]
             rows.append(row)
-
+
```
Comment (eggpi) [inline parse.py]: Looks like there's some extra space here.

File: config.py [hunk 12]
```diff
@@ -72,15 +41,63 @@
...
+        # Dictionary of section title to vector
+        section_path = os.path.expanduser('~/citation-needed/embeddings/section_dict_en.pck'),
+
+        # Tensorflow models to detect Citation Need for English
+        model_path = os.path.expanduser('~/citation-needed/models/fa_en_model_rnn_attention_section.h5'),
+
+        # Argument for padding word vectors to the same length 
+        # so as to use as the input for the RNN model
+        word_vector_length = 187,
+
```
Comment (eggpi) [inline config.py]: Nit: I'd remove this extra line.

File: config.py [hunk 13]
```diff
@@ -72,15 +41,63 @@
             'External links',
             'Further reading',
             'Notes',
+            'Additional sources',
+            'Sources',
```
Comment (eggpi) [inline config.py]: Since we have a dictionary of section names as part of the model, could we skip all sections that are not in the dictionary, rather than configuring here?

File: parse.py (hunk 5 above)
Comment (eggpi) [reply parse.py]: Wow, that's interesting. It seems a little bit odd to me that an unknown word can push known words out of the vector if it appears in the beginning of the sentence, but maybe that's by design -- I'll have to defer to @mirrys here.

File: config.py (hunk 13 above)
Comment (AikoChou) [reply config.py]: I think we need to keep them because `sections_to_skip` here specifies sections that we don't want to parse the contents and predict scores for sentences.

As I understand it, for sections that are not in the model dictionary `section_dict_en.pck` we may still want to parse contents and extract sentences to predict scores, although the section name we had never seen during the model training phase, so they are not in the dictionary.

**Comment by AikoChou** [comment]
Could I merge the pull request now? Will all the conversations retain after I merge?

File: config.py (hunk 13 above)
Comment (eggpi) [reply config.py]: Oh, yeah, I actually don't know if we can end up with predictions for sentences not in the dictionary or not. Another question for @mirrys I guess.

**Comment by eggpi** [comment]
The conversations will still exist in the pull request, but I worry we might forget about them if we merge it and continue development.

If you really want to merge it to master, maybe file issues for the open questions so we remember to follow up? I think these are the open ones now:

- https://github.com/AikoChou/citationdetective/pull/1#discussion_r363049322
- https://github.com/AikoChou/citationdetective/pull/1#discussion_r362445451
- https://github.com/AikoChou/citationdetective/pull/1#discussion_r362442619

I did ping Miriam about those too. We likely won't get answers over the weekend, though, so we could wait a little bit more.

File: config.py (hunk 13 above)
Comment (mirrys) [reply config.py]: Hi, as @AikoChou said, it would be better to keep them. The model will asign an 'unkown' value for unknown sections, and proceed in analysing the text of the sentence.

File: parse.py (hunk 5 above)
Comment (mirrys) [reply parse.py]: > I think because the input for the RNN model is a fixed length of the word vectors, if we don't remove extra hyperlinks, they become many `UNK` filling the word vectors. When the length exceeds the limitation of maximum word vectors, they are truncated and some meaningful words are missing.

@AikoChou is right, this is a limitation of the model design. But my understanding the maximum length for the fixed sentence has been calculated to be the largest sentence in the training set -- let me check with my co-author and get back to this.

More in general, because token are treated as words when present in the dictionary, or ar unknown when not recognized,  I would probably just include links as well and leave the word analysis process as close as possible to the original script.

File: utils.py (hunk 10 above)
Comment (mirrys) [reply utils.py]: I wouldn't advise it unless the function proposed is performing the same word processing as the one in the original script.

Comment (eggpi) [reply utils.py]: Yeah you're right, actually it would be a bad idea for us to change it here now. @AikoChou let's just leave this as-is for now please :)

For the future, it's not great to have to maintain a list of "cleanup regexes" (per language?!) in multiple places (in the scripts that train the model _and_ any clients that want to use it, such as Citation Detective). We should be thinking about stuff like:

- Moving this function to a library that is used for interacting with the model. This library is used for training new versions of the model and accessible to clients (such as Citation Detective) that want to use it, without having to duplicate code.
- Replacing these cleanup functions with tokenization/stemming/lemming functions from established Python packages where possible, so we don't have to maintain our own versions.

...though all of that is probably out of the scope of this Outreachy project.

File: parse.py (hunk 5 above)
Comment (AikoChou) [reply parse.py]: From my perspective, the links are unnecessary to keep because they belong to citations in this case, but we should keep the links if they show in the content.

File: utils.py (hunk 10 above)
Comment (AikoChou) [reply utils.py]: > I wouldn't advise it unless the function proposed is performing the same word processing as the one in the original script.

The 'word_tokenize' provided by nltk library is different from 'text_to_word_list' when processing contracted words like `can't` `would've` `don't` `they'd`. 'word_tokenize' does not convert them to original form. Also, when it comes to `12.1/km²`, 'word_tokenize' tends to preserve the whole but 'text_to_word_list' divides it as `12` `1` and `km`. 

I agree we should leave it not change for now because the model is trained based on the results from this function. But I feel this function may need to be improved as I found it does not remove punctuations like `;` `:` `-` and split words like `U.N.` and decimal numbers.

> For the future, it's not great to have to maintain a list of "cleanup regexes" (per language?!) in multiple places (in the scripts that train the model _and_ any clients that want to use it, such as Citation Detective). We should be thinking about stuff like:
> * Moving this function to a library that is used for interacting with the model. This library is used for training new versions of the model and accessible to clients (such as Citation Detective) that want to use it, without having to duplicate code.
> * Replacing these cleanup functions with tokenization/stemming/lemming functions from established Python packages where possible, so we don't have to maintain our own versions.

I will file an issue for this as future work in case we forget :)
